### PR TITLE
add fast failed in ReadWriteOncePod

### PR DESCRIPTION
**Pull request by boenn**
#### What type of PR is this?

/kind design

#### What this PR does / why we need it:

mark the pod `Unschedulable` and implement preemption logic in **volume_restrictions** plugin

#### Which issue(s) this PR fixes:

Fixes part of [#103132](https://github.com/kubernetes/kubernetes/issues/103132)

#### Special notes for your reviewer:

In the logic of preemption, I only pay attention to whether the PVC matches, and does not involve conditions such as affinity, taint/toleration, antiaffinity....

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig scheduling
/assign @chrishenzie

### CODE REVIEW

**Comment by chrishenzie** [comment]
/ok-to-test

Thank you for taking this on!

Some important context around this change and enforcing access modes in the scheduler in general:

- In API review it was suggested we either [use the scheduler to enforce both the `ReadWriteOnce` and `ReadWriteOncePod` access modes, or don't rely on the scheduler at all](https://github.com/kubernetes/kubernetes/pull/102028#issuecomment-870004026)
- [An issue was opened for enforcing the `ReadWriteOnce` access mode via the scheduler](https://github.com/kubernetes/kubernetes/issues/103305) to be consistent with `ReadWriteOncePod`
- However, [there is likely not a backwards compatible way of doing this for `ReadWriteOnce`](https://github.com/kubernetes/kubernetes/issues/103305#issuecomment-882720179) because of the behavior change it introduces

Given this and the API review feedback, we may want to roll back the scheduler changes for `ReadWriteOncePod` (to stay consistent with `ReadWriteOnce`) or re-evaluate if it is okay to allow this inconsistency.

/cc @msau42 
/cc @alculquicondor

**Comment by alculquicondor** [comment]
> Given this and the API review feedback, we may want to roll back the scheduler changes for ReadWriteOncePod

I would be against that. The inconsistency should be fine. In a v2 API (hopefully one day), we should fix ReadWriteOnce.

As for this PR, I only skimmed through it, but it seems to be missing some things that we discussed here: https://github.com/kubernetes/kubernetes/pull/103082#discussion_r656456477

**Comment by boenn** [comment]
> As for this PR, I only skimmed through it, but it seems to be missing some things that we discussed here: [ #103082 (comment)](https://github.com/kubernetes/kubernetes/pull/103082#discussion_r656456477)

Thanks for your reminder, I tried to modify the code, but it still looks a little weird.I think ReadWriteOncePod will achieve preemption in the future (ReadWriteOnce is the same), scheduler for both is right. 

The current implementation is that if pod1 and pod2 use pvc with the same name, the latter may preempt the former by default (of course, ReadWriteOncePod is used), which is quite rude. I checked [here](https://github.com/kubernetes/kubernetes/pull/102028#issuecomment-870004026) and thought we need to turn on a flag in ReadWriteOncePod to confirm whether preemption is allowed, which may be a lot more elegant. (This may solve the backward compatibility problem) or use some "private" path.

I'm not sure if I need to add an affinity/anti-affinity check here, it seems to be out of bounds..

**Comment by boenn** [comment]
I try to modify the logic to fail fast, or am I waiting for a conclusion [here](https://github.com/kubernetes/kubernetes/issues/103305) here and then modify it ... ? :worried:

**Comment by chrishenzie** [comment]
@boenn The only argument raised for not doing this is to be consistent with how we handle ReadWriteOnce. I think it's reasonable to keep iterating on this now, but we have a full release to land this feature so there isn't a rush. I think we will arrive at a consensus soon.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 1]
```diff
@@ -241,3 +291,81 @@ func New(_ runtime.Object, handle framework.Handle, fts feature.Features) (frame
...
+		for key := range pvcKeys {
+			refCount := nodeInfo.PVCRefCounts[key]
+			if refCount != 0 {
+				pvcKeysNumber--
+			}
+		}
+		if pvcKeysNumber == 0 {
+			s.nodes = append(s.nodes, nodeInfo.Node())
+		}
+		s.pvck = pvcKeys.Union(s.pvck)
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:352]: Aren't these keys independent of the pods running in the cluster? Initially, you just obtain them from the incoming Pod and the existing RWOP PVCs.

I'm stopping my review here. Please give the `preFilterState` variable names and comments that explain what they hold and how they change.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 2]
```diff
@@ -48,10 +48,14 @@ var _ framework.EnqueueExtensions = &VolumeRestrictions{}
 const Name = names.VolumeRestrictions
 
 const (
+	preFilterStateKey = "PreFilter" + Name
 	// ErrReasonDiskConflict is used for NoDiskConflict predicate error.
 	ErrReasonDiskConflict = "node(s) had no available disk"
 	// ErrReasonReadWriteOncePodConflict is used when a pod is found using the same PVC with the ReadWriteOncePod access mode.
-	ErrReasonReadWriteOncePodConflict = "node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode"
+	// ErrReasonReadWriteOncePodConflict = "node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode"
```
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: done.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 3]
```diff
@@ -48,10 +48,14 @@ var _ framework.EnqueueExtensions = &VolumeRestrictions{}
...
+
+	// ErrReasonReadWriteOncePodSame means the pod to be scheduled matches the pvcs of the pod on the node, which is unreasonable scheduling and is used for fast failure
+	ErrReasonReadWriteOncePodSame = "Schedule to this node for fast failure"
```
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: Attempt to solve explained the reason.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 4]
```diff
@@ -122,21 +126,84 @@ func haveOverlap(a1, a2 []string) bool {
...
+	}
+	pvcKeys, status := pl.GetPvcKeys(podToSchedule)
+	if status.IsUnschedulable() {
+		return status
+	}
+	state.updateWithPod(nodeInfo, pvcKeys, -1)
+	return nil
+}
+
+func (pl *VolumeRestrictions) GetPvcKeys(pod *v1.Pod) (sets.String, *framework.Status) {
```
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: done.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 5]
```diff
@@ -122,21 +126,84 @@ func haveOverlap(a1, a2 []string) bool {
...
+func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
 	nodeInfos, err := pl.nodeInfoLister.NodeInfos().List()
 	if err != nil {
 		return framework.NewStatus(framework.Error, "error while getting node info")
 	}
+	pvcKeys, status := pl.GetPvcKeys(pod)
+	if status != nil {
+		return status
+	}
+	nodes := make([]*v1.Node, 0)
```
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: fine, I modified the name, it can express some characteristics.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 6]
```diff
@@ -122,21 +126,84 @@ func haveOverlap(a1, a2 []string) bool {
...
+	processNode := func(i int) {
+		nodeInfo := nodeInfos[i]
+		pvcKeysNumber := len(pvcKeys)
```
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: Do you mean logic or naming? Yes, then match.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 7]
```diff
@@ -122,21 +126,84 @@ func haveOverlap(a1, a2 []string) bool {
...
+		for key := range pvcKeys {
+			refCount := nodeInfo.PVCRefCounts[key]
+			if refCount != 0 {
+				pvcKeysNumber--
+			}
+		}
+		if pvcKeysNumber == 0 {
+			nodes = append(nodes, nodeInfo.Node())
```
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: fine, I tried to modify..

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 8]
```diff
@@ -122,21 +126,84 @@ func haveOverlap(a1, a2 []string) bool {
...
+		}
+	}
+	pl.parallelizer.Until(context.Background(), len(nodeInfos), processNode)
+	// Enforce ReadWriteOncePod access mode. This is also enforced during volume mount in kubelet.
+	//if conflicts > 0 {
```
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: done

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 9]
```diff
@@ -122,21 +126,84 @@ func haveOverlap(a1, a2 []string) bool {
...
+	//	return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonReadWriteOncePodConflict)
+	//}
+
+	// update cyclestate
+	p := &preFilterState{
+		nodes: nodes,
```
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: Okay, I changed the name and added a comment

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 10]
```diff
@@ -122,21 +126,84 @@ func haveOverlap(a1, a2 []string) bool {
...
+		pvck:  pvcKeys,
```
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: done

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 11]
```diff
@@ -122,21 +126,84 @@ func haveOverlap(a1, a2 []string) bool {
...
+	}
+	cycleState.Write(preFilterStateKey, p)
+	return nil
+}
+
+func (pl *VolumeRestrictions) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
```
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: Umm..I tried to add an integration in `preemption_test.go`, but it seems to have some problems. I will try to find the reason..

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 12]
```diff
@@ -241,3 +291,81 @@ func New(_ runtime.Object, handle framework.Handle, fts feature.Features) (frame
...
+
+	s, ok := c.(*preFilterState)
+	if !ok {
+		return nil, fmt.Errorf("%+v  convert to volumerestrictions.state error", c)
+	}
+	return s, nil
+}
+
+func (s *preFilterState) updateWithPod(nodeInfo *framework.NodeInfo, pvcKeys sets.String, multiplier int64) {
+	if multiplier == 1 {
```
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: I am a little confused, it has been modified..

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go (hunk 1 above)
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: I tried to modify the naming of the fields in preFilterState and add notes.

By the way, if we only need to fail quickly, then the **keysWithRWOP** field is not needed. I am a bit confused about the subsequent preemption logic (such as the condition that triggers the preemption). If the final rwop handles preemption in the same way as rwo, I will delete this filed.

**Comment by boenn** [comment]
@chrishenzie Thank you for your reply. I am confused about the logic of preemption, it really needs a conclusion.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: Okay, I know ☂️

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 13]
```diff
@@ -122,21 +126,110 @@ func haveOverlap(a1, a2 []string) bool {
...
+			if state.nodesMatchPVC[i].Name == nodeInfo.Node().Name {
+				state.nodesMatchPVC = append(state.nodesMatchPVC[:i], state.nodesMatchPVC[i+1:]...)
+			}
+		}
+	}
+	state.keysWithRWOP = state.keysWithRWOP.Difference(pvcKeys)
+	return nil
+}
+
+// Collect a pod in the form of namespace+pvcName with pvc whose accessmode is ReadWriteOncePod.
```
Comment (chrishenzie) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:230]: It looks like this method is not collecting pods, but collecting the names of the PVCs on the pod with ReadWriteOncePod access mode

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 14]
```diff
@@ -48,10 +49,13 @@ var _ framework.EnqueueExtensions = &VolumeRestrictions{}
 const Name = names.VolumeRestrictions
 
 const (
+	preFilterStateKey = "PreFilter" + Name
 	// ErrReasonDiskConflict is used for NoDiskConflict predicate error.
 	ErrReasonDiskConflict = "node(s) had no available disk"
-	// ErrReasonReadWriteOncePodConflict is used when a pod is found using the same PVC with the ReadWriteOncePod access mode.
-	ErrReasonReadWriteOncePodConflict = "node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode"
```
Comment (chrishenzie) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:54]: What happens if two pods are started with the same priority and referencing the same PVC with ReadWriteOncePod? What error will be surfaced to users when one pod inevitably fails to schedule?

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 15]
```diff
@@ -122,21 +126,110 @@ func haveOverlap(a1, a2 []string) bool {
...
+				atomic.AddUint32(&pvcKeysNumber, ^uint32(0))
+				if pvcKeysNumber == 0 {
+					nodesMatchPVCWithRWOP = append(nodesMatchPVCWithRWOP, nodeInfo.Node())
+					cancel()
+				}
+			}
+		}
+	}
+	pl.parallelizer.Until(ctx, len(nodeInfos), processNode)
+	// update cyclestate
```
Comment (chrishenzie) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:165]: nit: Can you provide more detail here on why we need to store this information?

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 16]
```diff
@@ -122,21 +126,110 @@ func haveOverlap(a1, a2 []string) bool {
...
+	ctx, cancel := context.WithCancel(context.Background())
+	processNode := func(i int) {
+		nodeInfo := nodeInfos[i]
+		pvcKeysNumber := uint32(len(pvcKeys))
+		for key := range pvcKeys {
+			refCount := nodeInfo.PVCRefCounts[key]
+			if refCount != 0 {
+				atomic.AddUint32(&pvcKeysNumber, ^uint32(0))
+				if pvcKeysNumber == 0 {
+					nodesMatchPVCWithRWOP = append(nodesMatchPVCWithRWOP, nodeInfo.Node())
```
[the remark this answers is not in the document]
Comment (chrishenzie) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:158]: I think I see what's happening. We only want to append if the node has all of the PVCs with ReadWriteOncePod.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 17]
```diff
@@ -122,21 +126,110 @@ func haveOverlap(a1, a2 []string) bool {
...
 	if err != nil {
 		return framework.NewStatus(framework.Error, "error while getting node info")
 	}
+	pvcKeys, status := pl.keysForPVCWithRWOP(pod)
+	if status != nil {
+		return status
+	}
+	var nodesMatchPVCWithRWOP []*v1.Node
+	ctx, cancel := context.WithCancel(context.Background())
+	processNode := func(i int) {
```
Comment (chrishenzie) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:150]: nit: I think it might be worth adding a comment explaining what's happening here. From my understanding, we're collecting the nodes that contain all of the pod's PVCs

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 18]
```diff
@@ -241,3 +323,66 @@ func New(_ runtime.Object, handle framework.Handle, fts feature.Features) (frame
 		enableReadWriteOncePod: fts.EnableReadWriteOncePod,
 	}, nil
 }
+
+// canBePreempt only verifies the volume, not the resource/affinity..
+func canBePreempt(nodeInfo *framework.NodeInfo, key string) bool {
```
Comment (chrishenzie) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:328]: How does the preemption logic account for pod priorities?

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 19]
```diff
@@ -48,10 +49,13 @@ var _ framework.EnqueueExtensions = &VolumeRestrictions{}
...
 const (
+	preFilterStateKey = "PreFilter" + Name
 	// ErrReasonDiskConflict is used for NoDiskConflict predicate error.
 	ErrReasonDiskConflict = "node(s) had no available disk"
-	// ErrReasonReadWriteOncePodConflict is used when a pod is found using the same PVC with the ReadWriteOncePod access mode.
-	ErrReasonReadWriteOncePodConflict = "node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode"
+	// ErrReasonReadWriteOncePodRepeat means that after the ReadWriteOncePod is turned on,
+	// the current cluster has nodes that meet the volume restrictions,
+	// so the pod will be scheduled to that node instead of this node.
+	ErrReasonReadWriteOncePodRepeat = "pod will be scheduled to nodes that meet volume restrictions"
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:58]: why did you change this?

This message should explain why the pod didn't schedule. This is not doing that.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 20]
```diff
@@ -30,6 +29,8 @@ import (
 	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/feature"
 	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/names"
 	"k8s.io/kubernetes/pkg/scheduler/internal/parallelize"
+	"strings"
+	"sync/atomic"
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:33]: these 2 imports should go in the first block of imports

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 21]
```diff
@@ -122,21 +126,110 @@ func haveOverlap(a1, a2 []string) bool {
...
+	}
+	var nodesMatchPVCWithRWOP []*v1.Node
+	ctx, cancel := context.WithCancel(context.Background())
+	processNode := func(i int) {
+		nodeInfo := nodeInfos[i]
+		pvcKeysNumber := uint32(len(pvcKeys))
+		for key := range pvcKeys {
+			refCount := nodeInfo.PVCRefCounts[key]
+			if refCount != 0 {
+				atomic.AddUint32(&pvcKeysNumber, ^uint32(0))
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:156]: `pvcKeysNumber` is a local variable to the thread, you don't need atomic operations for it.

The problem is the `append` below

File: test/integration/scheduler/util.go [hunk 22]
```diff
@@ -19,6 +19,11 @@ package scheduler
 import (
 	"context"
 	"fmt"
+	"k8s.io/apimachinery/pkg/watch"
```
Comment (alculquicondor) [inline test/integration/scheduler/util.go:22]: all these imports should go in the import block below

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go (hunk 18 above)
Comment (alculquicondor) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:328]: this is unused, right?

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 23]
```diff
@@ -241,3 +323,66 @@ func New(_ runtime.Object, handle framework.Handle, fts feature.Features) (frame
...
+}
+
+// preFilterState computed at PreFilter and used at Filter.
+// PreFilter traverses and compares whether the pvc referenced by each pod on each node
+// and the pvc of podToScheduled are the same pvc
+type preFilterState struct {
+	// All nodes that can match podToScheduled's all pvcs
+	// Used for fast-failure scheduling
+	nodesMatchPVC []*v1.Node
+	// If the AccessModes of the pvc used by a pod is ReadWriteOncePod,
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:352]: is it pods or PVCs? who do they belong to? the incoming Pod or all the Pods in the cluster?

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go [hunk 24]
```diff
@@ -241,3 +323,66 @@ func New(_ runtime.Object, handle framework.Handle, fts feature.Features) (frame
...
+		}
+	}
+	return false
+}
+
+// preFilterState computed at PreFilter and used at Filter.
+// PreFilter traverses and compares whether the pvc referenced by each pod on each node
+// and the pvc of podToScheduled are the same pvc
+type preFilterState struct {
+	// All nodes that can match podToScheduled's all pvcs
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:349]: What do you mean by "match"? What is the condition that the nodes in this list satisfy?

**Comment by chrishenzie** [comment]
Hello @boenn !

It looks like this problem has complexity that requires a contributor with past experience with storage and/or preemption.

I welcome you to take a look for any issues labeled as [`good first issue`](https://github.com/kubernetes/kubernetes/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) or [`help wanted`](https://github.com/kubernetes/kubernetes/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22). It would be great to have you review my PR if I can get to fixing this soon. Thank you!

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:158]: > I think I see what's happening. We only want to append if the node has all of the PVCs with ReadWriteOncePod.

Yes, I don’t know whether it should be limited to referencing all PVCs or referencing some PVC nodes

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go (hunk 15 above)
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:165]: > nit: Can you provide more detail here on why we need to store this information or how it will be used later?

https://github.com/kubernetes/kubernetes/blob/acfdb93f752a3635f41aa177944b81b8512f3cb2/pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go#L345-L355

I have explained the fields here.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go (hunk 13 above)
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:230]: Sorry, I was careless

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go (hunk 18 above)
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:328]: Sorry I will delete it. Preemption is neither used nor implemented, now only the scheduling problem is considered

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go (hunk 24 above)
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:349]: If the pod of a node references the pvc (rwop) of all pods to be scheduled, this node will be added.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go (hunk 23 above)
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:352]: incoming Pod.The comment has been updated.

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go (hunk 24 above)
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:349]: I tried to fix it

File: pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go (hunk 21 above)
Comment (boenn) [reply pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go]: I tried to fix it

**Comment by boenn** [comment]
@chrishenzie Thank you for your suggestion. What I am a little confused about is whether we should wait for results before making changes. Now it is a bit confusing. Regardless of which nodes are scheduled to or preemption issues, it has not yet been determined.

**Comment by MadhavJivrajani** [comment]
/remove-kind design
/kind feature
`/kind design` is being deprecated in favor of `/kind feature`
https://github.com/kubernetes/community/issues/5641

**Comment by chrishenzie** [comment]
I believe we can close this, we are tackling in a separate PR (still in draft state)
